File: src/check.rs
use std::{
    collections::{BTreeMap, BTreeSet},
    fs::File,
    io::BufReader,
    process::{Command, Output, Stdio},
    vec,
};

use crate::flake::{Input, Locks, Metadata, Node};
use serde::de::DeserializeOwned;

use crate::{
    config::Config,
    error::{Error, ProcessInfo},
    options::CheckOptions,
};

const CONFIG_ATTRIBUTE_PATH: &str = "flatFlake";

pub fn check(check_options: CheckOptions) -> Result<(), Error> {
    let config = get_config(&check_options)?;
    log::debug!("config: {:#?}", config);

    let locks = get_locks(&check_options)?;
    log::debug!("{:#?}", locks);

    let Locks { root, nodes, .. } = &locks;

    let root_node = &get_node(nodes, root)?;

    let mut allowed = config.allowed;
    allowed.extend(root_node.inputs.iter().filter_map(|(k, i)| match i {
        Input::Introduce(_) => Some(vec![k.to_owned()]),
        Input::Follow(_) => None,
    }));

    let mut not_allowed = BTreeSet::new();

    let mut path: Vec<String> = vec![];
    check_node(nodes, &mut allowed, root_node, &mut path, &mut not_allowed)?;

    if !not_allowed.is_empty() {
        return Err(Error::NotFlat(not_allowed));
    }

    if !allowed.is_empty() {
        return Err(Error::UnusedAllowed(allowed));
    }

    Ok(())
}

pub fn get_config(options: &CheckOptions) -> Result<Config, Error> {
    if let Some(path) = &options.config_file {
        let file = BufReader::new(File::open(path)?);
        return Ok(serde_json::from_reader(file)?);
    }

    let flake = &options.flake;
    match call_nix(["eval", &format!("{flake}#{CONFIG_ATTRIBUTE_PATH}")]) {
        Ok(c) => Ok(c),
        Err(e) => {
            log::debug!("error: {}", e);
            log::info!(
                "failed to eval '{flake}#{CONFIG_ATTRIBUTE_PATH}', use default configuration"
            );
            Ok(Default::default())
        }
    }
}

pub fn get_locks(options: &CheckOptions) -> Result<Locks, Error> {
    if let Some(path) = &options.lock_file {
        let file = BufReader::new(File::open(path)?);
        return Ok(serde_json::from_reader(file)?);
    }

    let metadata: Metadata = call_nix(["flake", "metadata", &options.flake])?;
    Ok(metadata.locks)
}

pub fn check_node(
    nodes: &BTreeMap<String, Node>,
    allowed: &mut BTreeSet<Vec<String>>,
    current: &Node,
    path: &mut Vec<String>,
    not_allowed: &mut BTreeSet<Vec<String>>,
) -> Result<(), Error> {
    log::debug!("checking {path:?}");
    for (name, input) in &current.inputs {
        path.push(name.to_owned());
        if let Input::Introduce(key) = input {
            if !allowed.remove(&*path) {
                not_allowed.insert(path.clone());
            }
            let next = get_node(nodes, key)?;
            check_node(nodes, allowed, next, path, not_allowed)?;
        }
        path.pop();
    }
    Ok(())
}

pub fn get_node<'a>(nodes: &'a BTreeMap<String, Node>, key: &str) -> Result<&'a Node, Error> {
    nodes
        .get(key)
        .ok_or_else(|| Error::MissingLockNode(key.to_string()))
}

pub fn call_nix<I, S, T>(args: I) -> Result<T, Error>
where
    I: IntoIterator<Item = S>,
    S: AsRef<str>,
    T: DeserializeOwned,
{
    let mut args_vec: Vec<_> = args.into_iter().map(|s| s.as_ref().to_owned()).collect();
    args_vec.extend(
        [
            "--json",
            "--extra-experimental-features",
            "nix-command flakes",
        ]
        .into_iter()
        .map(|s| s.to_owned()),
    );
    let command_vec = {
        let mut cmd = vec!["nix".to_owned()];
        cmd.extend(args_vec.iter().cloned());
        cmd
    };

    let mut command = Command::new("nix");
    command.args(args_vec);

    log::trace!("run command: {:?}", command_vec);

    let output = get_command_output(command)?;

    let process_info = ProcessInfo {
        command: command_vec,
        status: output.status,
        stdout: String::from_utf8(output.stdout)?,
        stderr: String::from_utf8(output.stderr)?,
    };
    log::trace!("command called:\n{}", process_info);
    if !output.status.success() {
        return Err(Error::ProcessFailed(process_info));
    }

    Ok(serde_json::from_str(&process_info.stdout)?)
}

pub fn get_command_output(mut command: Command) -> Result<Output, Error> {
    let handle = command
        .stdout(Stdio::piped())
        .stderr(Stdio::inherit())
        .spawn()?;
    Ok(handle.wait_with_output()?)
}

File: src2/check.ts
import { FlatFlakeError, ProcessFailedError, WrappedError } from "@/error.ts";
import { parseConfig, type Config } from "@/config.ts";
import { getNode, type Locks, type Node } from "@/flake.ts";

export interface CheckOptions {
  flake: string;
  lockFile?: string;
  configFile?: string;
}

const CONFIG_ATTRIBUTE_PATH = "flatFlake";

/** Outcome of a check: the two "expected" failures are plain return values. */
export interface CheckResult {
  ok: boolean;
  notAllowed: string[][];
  unused: string[][];
}

export async function check(options: CheckOptions): Promise<CheckResult> {
  const config = await getConfig(options);

  const locks = await getLocks(options);

  const { root, nodes } = locks;
  const rootNode = getNode(nodes, root);

  // Top-level inputs are implicitly allowed.
  const allowed: string[][] = [...config.allowed];
  for (const [name, input] of Object.entries(rootNode.inputs ?? {})) {
    if (typeof input === "string") allowed.push([name]);
  }

  const notAllowed: string[][] = [];
  checkNode(nodes, allowed, rootNode, [], notAllowed);

  return {
    ok: notAllowed.length === 0 && allowed.length === 0,
    notAllowed,
    unused: allowed,
  };
}

/** Recursively verify that every introduced input is explicitly allowed. */
export function checkNode(
  nodes: Record<string, Node>,
  allowed: string[][],
  current: Node,
  path: string[],
  notAllowed: string[][],
): void {
  for (const [name, input] of Object.entries(current.inputs ?? {})) {
    path.push(name);
    if (typeof input === "string") {
      const index = allowed.findIndex((entry) => samePath(entry, path));
      if (index < 0) {
        notAllowed.push([...path]);
      } else {
        allowed.splice(index, 1);
      }
      const next = getNode(nodes, input);
      checkNode(nodes, allowed, next, path, notAllowed);
    }
    // `follows` (array) inputs introduce nothing new and are skipped.
    path.pop();
  }
}

/** Input paths are visited exactly once, so reference-free equality suffices. */
function samePath(a: string[], b: string[]): boolean {
  return a.length === b.length && a.every((component, index) => component === b[index]);
}

async function getConfig(options: CheckOptions): Promise<Config> {
  if (options.configFile) {
    return readJsonFile(options.configFile).then(parseConfig);
  }

  const flake = options.flake;
  try {
    const raw = await callNix(["eval", `${flake}#${CONFIG_ATTRIBUTE_PATH}`]);
    return parseConfig(raw);
  } catch (error) {
    console.warn(`${Bun.color("yellow", "ansi")}[WARN] failed to eval '${flake}#${CONFIG_ATTRIBUTE_PATH}', use default configuration`);
    return { allowed: [] };
  }
}

async function getLocks(options: CheckOptions): Promise<Locks> {
  if (options.lockFile) {
    // A lock file contains the `locks` object at the top level.
    return parseLocks(await readJsonFile(options.lockFile));
  }

  const metadata = await callNix(["flake", "metadata", options.flake]);
  if (metadata === null || typeof metadata !== "object" || !("locks" in metadata)) {
    throw new FlatFlakeError("flake metadata did not contain a 'locks' field");
  }
  return parseLocks((metadata as Record<string, unknown>).locks);
}

/** Validate a `flake.lock` object (root + nodes) at runtime. */
function parseLocks(raw: unknown): Locks {
  if (raw === null || typeof raw !== "object") {
    throw new FlatFlakeError("invalid flake lock: expected a JSON object");
  }
  const { root, nodes } = raw as Record<string, unknown>;
  if (typeof root !== "string" || root.length === 0) {
    throw new FlatFlakeError("flake lock is missing a root node");
  }
  if (nodes === null || typeof nodes !== "object") {
    throw new FlatFlakeError("flake lock is missing its 'nodes' map");
  }
  return raw as Locks;
}

async function readJsonFile(path: string): Promise<unknown> {
  try {
    return await Bun.file(path).json();
  } catch (error) {
    throw new WrappedError(error, `failed to read ${path}`);
  }
}

/** Run `nix <args> --json --extra-experimental-features 'nix-command flakes'`. */
export async function callNix<T = unknown>(args: string[]): Promise<T> {
  const command = [
    "nix",
    ...args,
    "--json",
    "--extra-experimental-features",
    "nix-command flakes",
  ];

  const result = Bun.spawnSync({
    cmd: command,
    stdout: "pipe",
    stderr: "inherit",
  });
  const stdout = new TextDecoder().decode(result.stdout);

  if (!result.success) {
    throw new ProcessFailedError({
      command,
      status: result.exitCode,
      signal: result.signalCode ?? null,
      stdout,
    });
  }

  try {
    return JSON.parse(stdout) as T;
  } catch (error) {
    throw new WrappedError(error, "failed to parse nix output as JSON");
  }
}
